// @trigger.dev/core/dist/commonjs/v3/task-catalog/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TaskCatalogAPI = void 0;
const API_NAME = "task-catalog";
const globals_js_1 = require("../utils/globals.js");
const noopTaskCatalog_js_1 = require("./noopTaskCatalog.js");
const NOOP_TASK_CATALOG = new noopTaskCatalog_js_1.NoopTaskCatalog();
class TaskCatalogAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new TaskCatalogAPI();
        }
        return this._instance;
    }
    setGlobalTaskCatalog(taskCatalog) {
        return (0, globals_js_1.registerGlobal)(API_NAME, taskCatalog);
    }
    disable() {
        (0, globals_js_1.unregisterGlobal)(API_NAME);
    }
    registerTaskMetadata(task) {
        this.#getCatalog().registerTaskMetadata(task);
    }
    updateTaskMetadata(id, updates) {
        this.#getCatalog().updateTaskMetadata(id, updates);
    }
    registerTaskFileMetadata(id, metadata) {
        this.#getCatalog().registerTaskFileMetadata(id, metadata);
    }
    listTaskManifests() {
        return this.#getCatalog().listTaskManifests();
    }
    getTaskManifest(id) {
        return this.#getCatalog().getTaskManifest(id);
    }
    getTask(id) {
        return this.#getCatalog().getTask(id);
    }
    taskExists(id) {
        return this.#getCatalog().taskExists(id);
    }
    #getCatalog() {
        return (0, globals_js_1.getGlobal)(API_NAME) ?? NOOP_TASK_CATALOG;
    }
}
exports.TaskCatalogAPI = TaskCatalogAPI;
