// @trigger.dev/core/dist/commonjs/v3/logger/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoggerAPI = void 0;
const taskLogger_js_1 = require("./taskLogger.js");
const globals_js_1 = require("../utils/globals.js");
const API_NAME = "logger";
const NOOP_TASK_LOGGER = new taskLogger_js_1.NoopTaskLogger();
class LoggerAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new LoggerAPI();
        }
        return this._instance;
    }
    disable() {
        (0, globals_js_1.unregisterGlobal)(API_NAME);
    }
    setGlobalTaskLogger(taskLogger) {
        return (0, globals_js_1.registerGlobal)(API_NAME, taskLogger);
    }
    debug(message, metadata) {
        this.#getTaskLogger().debug(message, metadata);
    }
    log(message, metadata) {
        this.#getTaskLogger().log(message, metadata);
    }
    info(message, metadata) {
        this.#getTaskLogger().info(message, metadata);
    }
    warn(message, metadata) {
        this.#getTaskLogger().warn(message, metadata);
    }
    error(message, metadata) {
        this.#getTaskLogger().error(message, metadata);
    }
    trace(name, fn) {
        return this.#getTaskLogger().trace(name, fn);
    }
    #getTaskLogger() {
        return (0, globals_js_1.getGlobal)(API_NAME) ?? NOOP_TASK_LOGGER;
    }
}
exports.LoggerAPI = LoggerAPI;
