// @trigger.dev/core/dist/commonjs/v3/apiClientManager/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.APIClientManagerAPI = void 0;
const index_js_1 = require("../apiClient/index.js");
const globals_js_1 = require("../utils/globals.js");
const getEnv_js_1 = require("../utils/getEnv.js");
const API_NAME = "api-client";
class APIClientManagerAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new APIClientManagerAPI();
        }
        return this._instance;
    }
    disable() {
        (0, globals_js_1.unregisterGlobal)(API_NAME);
    }
    setGlobalAPIClientConfiguration(config) {
        return (0, globals_js_1.registerGlobal)(API_NAME, config);
    }
    get baseURL() {
        const store = this.#getConfig();
        return store?.baseURL ?? (0, getEnv_js_1.getEnvVar)("TRIGGER_API_URL") ?? "https://api.trigger.dev";
    }
    get accessToken() {
        const store = this.#getConfig();
        return store?.secretKey ?? (0, getEnv_js_1.getEnvVar)("TRIGGER_SECRET_KEY") ?? (0, getEnv_js_1.getEnvVar)("TRIGGER_ACCESS_TOKEN");
    }
    get client() {
        if (!this.baseURL || !this.accessToken) {
            return undefined;
        }
        return new index_js_1.ApiClient(this.baseURL, this.accessToken);
    }
    #getConfig() {
        return (0, globals_js_1.getGlobal)(API_NAME);
    }
}
exports.APIClientManagerAPI = APIClientManagerAPI;
