// @trigger.dev/sdk/dist/commonjs/v3/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.envvars = exports.schedules = exports.runs = exports.logger = exports.AbortTaskRunError = exports.UnprocessableEntityError = exports.RateLimitError = exports.PermissionDeniedError = exports.NotFoundError = exports.InternalServerError = exports.ConflictError = exports.BadRequestError = exports.AuthenticationError = exports.ApiError = exports.queue = exports.retry = void 0;
exports.configure = configure;
__exportStar(require("./cache.js"), exports);
__exportStar(require("./config.js"), exports);
var retry_js_1 = require("./retry.js");
Object.defineProperty(exports, "retry", { enumerable: true, get: function () { return retry_js_1.retry; } });
var shared_js_1 = require("./shared.js");
Object.defineProperty(exports, "queue", { enumerable: true, get: function () { return shared_js_1.queue; } });
__exportStar(require("./tasks.js"), exports);
__exportStar(require("./wait.js"), exports);
__exportStar(require("./usage.js"), exports);
__exportStar(require("./idempotencyKeys.js"), exports);
__exportStar(require("./tags.js"), exports);
const v3_1 = require("@trigger.dev/core/v3");
var v3_2 = require("@trigger.dev/core/v3");
Object.defineProperty(exports, "ApiError", { enumerable: true, get: function () { return v3_2.ApiError; } });
Object.defineProperty(exports, "AuthenticationError", { enumerable: true, get: function () { return v3_2.AuthenticationError; } });
Object.defineProperty(exports, "BadRequestError", { enumerable: true, get: function () { return v3_2.BadRequestError; } });
Object.defineProperty(exports, "ConflictError", { enumerable: true, get: function () { return v3_2.ConflictError; } });
Object.defineProperty(exports, "InternalServerError", { enumerable: true, get: function () { return v3_2.InternalServerError; } });
Object.defineProperty(exports, "NotFoundError", { enumerable: true, get: function () { return v3_2.NotFoundError; } });
Object.defineProperty(exports, "PermissionDeniedError", { enumerable: true, get: function () { return v3_2.PermissionDeniedError; } });
Object.defineProperty(exports, "RateLimitError", { enumerable: true, get: function () { return v3_2.RateLimitError; } });
Object.defineProperty(exports, "UnprocessableEntityError", { enumerable: true, get: function () { return v3_2.UnprocessableEntityError; } });
Object.defineProperty(exports, "AbortTaskRunError", { enumerable: true, get: function () { return v3_2.AbortTaskRunError; } });
Object.defineProperty(exports, "logger", { enumerable: true, get: function () { return v3_2.logger; } });
var runs_js_1 = require("./runs.js");
Object.defineProperty(exports, "runs", { enumerable: true, get: function () { return runs_js_1.runs; } });
exports.schedules = __importStar(require("./schedules/index.js"));
exports.envvars = __importStar(require("./envvars.js"));
/**
 * Register the global API client configuration. Alternatively, you can set the `TRIGGER_SECRET_KEY` and `TRIGGER_API_URL` environment variables.
 * @param options The API client configuration.
 * @param options.baseURL The base URL of the Trigger API. (default: `https://api.trigger.dev`)
 * @param options.secretKey The secret key to authenticate with the Trigger API. (default: `process.env.TRIGGER_SECRET_KEY`) This can be found in your Trigger.dev project "API Keys" settings.
 *
 * @example
 *
 * ```typescript
 * import { configure } from "@trigger.dev/sdk/v3";
 *
 * configure({
 *  baseURL: "https://api.trigger.dev",
 *  secretKey: "tr_dev_1234567890"
 * });
 * ```
 */
function configure(options) {
    v3_1.apiClientManager.setGlobalAPIClientConfiguration(options);
}
