// @trigger.dev/sdk/dist/commonjs/triggers/notifications.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MissingConnectionResolvedNotification = exports.MissingConnectionNotification = void 0;
exports.missingConnectionNotification = missingConnectionNotification;
exports.missingConnectionResolvedNotification = missingConnectionResolvedNotification;
const core_1 = require("@trigger.dev/core");
function missingConnectionNotification(integrations) {
    return new MissingConnectionNotification({ integrations });
}
function missingConnectionResolvedNotification(integrations) {
    return new MissingConnectionResolvedNotification({ integrations });
}
class MissingConnectionNotification {
    options;
    constructor(options) {
        this.options = options;
    }
    get event() {
        return {
            name: core_1.MISSING_CONNECTION_NOTIFICATION,
            title: "Missing Connection Notification",
            source: "trigger.dev",
            icon: "connection-alert",
            parsePayload: core_1.MissingConnectionNotificationPayloadSchema.parse,
            properties: [
                {
                    label: "Integrations",
                    text: this.options.integrations.map((i) => i.id).join(", "),
                },
            ],
        };
    }
    attachToJob(triggerClient, job) { }
    get preprocessRuns() {
        return false;
    }
    async verifyPayload(payload) {
        return { success: true };
    }
    toJSON() {
        return {
            type: "static",
            title: this.event.title,
            rule: {
                event: this.event.name,
                source: "trigger.dev",
                payload: {
                    client: {
                        id: this.options.integrations.map((i) => i.id),
                    },
                },
            },
        };
    }
}
exports.MissingConnectionNotification = MissingConnectionNotification;
class MissingConnectionResolvedNotification {
    options;
    constructor(options) {
        this.options = options;
    }
    get event() {
        return {
            name: core_1.MISSING_CONNECTION_RESOLVED_NOTIFICATION,
            title: "Missing Connection Resolved Notification",
            source: "trigger.dev",
            icon: "connection-alert",
            parsePayload: core_1.MissingConnectionResolvedNotificationPayloadSchema.parse,
            properties: [
                {
                    label: "Integrations",
                    text: this.options.integrations.map((i) => i.id).join(", "),
                },
            ],
        };
    }
    attachToJob(triggerClient, job) { }
    get preprocessRuns() {
        return false;
    }
    async verifyPayload(payload) {
        return { success: true };
    }
    toJSON() {
        return {
            type: "static",
            title: this.event.title,
            rule: {
                event: this.event.name,
                source: "trigger.dev",
                payload: {
                    client: {
                        id: this.options.integrations.map((i) => i.id),
                    },
                },
            },
        };
    }
}
exports.MissingConnectionResolvedNotification = MissingConnectionResolvedNotification;
