// @trigger.dev/core/dist/esm/v3/apiClientManager/index.js
import { ApiClient } from "../apiClient/index.js";
import { getGlobal, registerGlobal, unregisterGlobal } from "../utils/globals.js";
import { getEnvVar } from "../utils/getEnv.js";
const API_NAME = "api-client";
export class APIClientManagerAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new APIClientManagerAPI();
        }
        return this._instance;
    }
    disable() {
        unregisterGlobal(API_NAME);
    }
    setGlobalAPIClientConfiguration(config) {
        return registerGlobal(API_NAME, config);
    }
    get baseURL() {
        const store = this.#getConfig();
        return store?.baseURL ?? getEnvVar("TRIGGER_API_URL") ?? "https://api.trigger.dev";
    }
    get accessToken() {
        const store = this.#getConfig();
        return store?.secretKey ?? getEnvVar("TRIGGER_SECRET_KEY") ?? getEnvVar("TRIGGER_ACCESS_TOKEN");
    }
    get client() {
        if (!this.baseURL || !this.accessToken) {
            return undefined;
        }
        return new ApiClient(this.baseURL, this.accessToken);
    }
    #getConfig() {
        return getGlobal(API_NAME);
    }
}
