// @trigger.dev/core/dist/commonjs/v3/usage/api.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsageAPI = void 0;
const API_NAME = "usage";
const globals_js_1 = require("../utils/globals.js");
const noopUsageManager_js_1 = require("./noopUsageManager.js");
const NOOP_USAGE_MANAGER = new noopUsageManager_js_1.NoopUsageManager();
class UsageAPI {
    static _instance;
    constructor() { }
    static getInstance() {
        if (!this._instance) {
            this._instance = new UsageAPI();
        }
        return this._instance;
    }
    setGlobalUsageManager(manager) {
        return (0, globals_js_1.registerGlobal)(API_NAME, manager);
    }
    disable() {
        this.#getUsageManager().disable();
        (0, globals_js_1.unregisterGlobal)(API_NAME);
    }
    start() {
        return this.#getUsageManager().start();
    }
    stop(measurement) {
        return this.#getUsageManager().stop(measurement);
    }
    pauseAsync(cb) {
        return this.#getUsageManager().pauseAsync(cb);
    }
    sample() {
        return this.#getUsageManager().sample();
    }
    #getUsageManager() {
        return (0, globals_js_1.getGlobal)(API_NAME) ?? NOOP_USAGE_MANAGER;
    }
}
exports.UsageAPI = UsageAPI;
