// @trigger.dev/core/dist/commonjs/schemas/events.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CancelRunsForEventSchema = exports.GetEventSchema = void 0;
const zod_1 = require("zod");
const runs_js_1 = require("./runs.js");
exports.GetEventSchema = zod_1.z.object({
    /** The event id */
    id: zod_1.z.string(),
    /** The event name */
    name: zod_1.z.string(),
    /** When the event was created */
    createdAt: zod_1.z.coerce.date(),
    /** When the event was last updated */
    updatedAt: zod_1.z.coerce.date(),
    /** The runs that were triggered by the event */
    runs: zod_1.z.array(zod_1.z.object({
        /** The Run id */
        id: zod_1.z.string(),
        /** The Run status */
        status: runs_js_1.RunStatusSchema,
        /** When the run started */
        startedAt: zod_1.z.coerce.date().optional().nullable(),
        /** When the run completed */
        completedAt: zod_1.z.coerce.date().optional().nullable(),
    })),
});
exports.CancelRunsForEventSchema = zod_1.z.object({
    cancelledRunIds: zod_1.z.array(zod_1.z.string()),
    failedToCancelRunIds: zod_1.z.array(zod_1.z.string()),
});
