// @trigger.dev/core/dist/commonjs/schemas/notifications.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MissingConnectionResolvedNotificationPayloadSchema = exports.MissingExternalConnectionResolvedNotificationPayloadSchema = exports.MissingDeveloperConnectionResolvedNotificationPayloadSchema = exports.CommonMissingConnectionNotificationResolvedPayloadSchema = exports.MissingConnectionNotificationPayloadSchema = exports.MissingExternalConnectionNotificationPayloadSchema = exports.MissingDeveloperConnectionNotificationPayloadSchema = exports.CommonMissingConnectionNotificationPayloadSchema = exports.MISSING_CONNECTION_RESOLVED_NOTIFICATION = exports.MISSING_CONNECTION_NOTIFICATION = void 0;
const zod_1 = require("zod");
exports.MISSING_CONNECTION_NOTIFICATION = "dev.trigger.notifications.missingConnection";
exports.MISSING_CONNECTION_RESOLVED_NOTIFICATION = "dev.trigger.notifications.missingConnectionResolved";
exports.CommonMissingConnectionNotificationPayloadSchema = zod_1.z.object({
    id: zod_1.z.string(),
    client: zod_1.z.object({
        id: zod_1.z.string(),
        title: zod_1.z.string(),
        scopes: zod_1.z.array(zod_1.z.string()),
        createdAt: zod_1.z.coerce.date(),
        updatedAt: zod_1.z.coerce.date(),
    }),
    authorizationUrl: zod_1.z.string(),
});
exports.MissingDeveloperConnectionNotificationPayloadSchema = exports.CommonMissingConnectionNotificationPayloadSchema.extend({
    type: zod_1.z.literal("DEVELOPER"),
});
exports.MissingExternalConnectionNotificationPayloadSchema = exports.CommonMissingConnectionNotificationPayloadSchema.extend({
    type: zod_1.z.literal("EXTERNAL"),
    account: zod_1.z.object({
        id: zod_1.z.string(),
        metadata: zod_1.z.any(),
    }),
});
exports.MissingConnectionNotificationPayloadSchema = zod_1.z.discriminatedUnion("type", [
    exports.MissingDeveloperConnectionNotificationPayloadSchema,
    exports.MissingExternalConnectionNotificationPayloadSchema,
]);
exports.CommonMissingConnectionNotificationResolvedPayloadSchema = zod_1.z.object({
    id: zod_1.z.string(),
    client: zod_1.z.object({
        id: zod_1.z.string(),
        title: zod_1.z.string(),
        scopes: zod_1.z.array(zod_1.z.string()),
        createdAt: zod_1.z.coerce.date(),
        updatedAt: zod_1.z.coerce.date(),
        integrationIdentifier: zod_1.z.string(),
        integrationAuthMethod: zod_1.z.string(),
    }),
    expiresAt: zod_1.z.coerce.date(),
});
exports.MissingDeveloperConnectionResolvedNotificationPayloadSchema = exports.CommonMissingConnectionNotificationResolvedPayloadSchema.extend({
    type: zod_1.z.literal("DEVELOPER"),
});
exports.MissingExternalConnectionResolvedNotificationPayloadSchema = exports.CommonMissingConnectionNotificationResolvedPayloadSchema.extend({
    type: zod_1.z.literal("EXTERNAL"),
    account: zod_1.z.object({
        id: zod_1.z.string(),
        metadata: zod_1.z.any(),
    }),
});
exports.MissingConnectionResolvedNotificationPayloadSchema = zod_1.z.discriminatedUnion("type", [
    exports.MissingDeveloperConnectionResolvedNotificationPayloadSchema,
    exports.MissingExternalConnectionResolvedNotificationPayloadSchema,
]);
