// @trigger.dev/core/dist/commonjs/schemas/statuses.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JobRunStatusRecordSchema = exports.StatusHistorySchema = exports.StatusUpdateSchema = exports.StatusUpdateStateSchema = void 0;
const zod_1 = require("zod");
const json_js_1 = require("./json.js");
exports.StatusUpdateStateSchema = zod_1.z.union([
    zod_1.z.literal("loading"),
    zod_1.z.literal("success"),
    zod_1.z.literal("failure"),
]);
const StatusUpdateDataSchema = zod_1.z.record(json_js_1.SerializableJsonSchema);
exports.StatusUpdateSchema = zod_1.z.object({
    label: zod_1.z.string().optional(),
    state: exports.StatusUpdateStateSchema.optional(),
    data: StatusUpdateDataSchema.optional(),
});
const InitalStatusUpdateSchema = exports.StatusUpdateSchema.required({ label: true });
exports.StatusHistorySchema = zod_1.z.array(exports.StatusUpdateSchema);
exports.JobRunStatusRecordSchema = InitalStatusUpdateSchema.extend({
    key: zod_1.z.string(),
    history: exports.StatusHistorySchema,
});
