// @trigger.dev/sdk/dist/esm/triggers/eventTrigger.js
import { deepMergeFilters } from "@trigger.dev/core";
import { formatSchemaErrors } from "../utils/formatSchemaErrors.js";
import { ParsedPayloadSchemaError } from "../errors.js";
export class EventTrigger {
    #options;
    constructor(options) {
        this.#options = options;
    }
    toJSON() {
        return {
            type: "static",
            title: this.#options.name ?? this.#options.event.title,
            rule: {
                event: this.#options.name ?? this.#options.event.name,
                source: this.#options.source ?? "trigger.dev",
                payload: deepMergeFilters(this.#options.filter ?? {}, this.#options.event.filter ?? {}),
            },
        };
    }
    get event() {
        return this.#options.event;
    }
    attachToJob(triggerClient, job) { }
    get preprocessRuns() {
        return false;
    }
    async verifyPayload(payload) {
        if (this.#options.verify) {
            if (payload instanceof Request) {
                const clonedRequest = payload.clone();
                return this.#options.verify(clonedRequest);
            }
        }
        return { success: true };
    }
}
/** `eventTrigger()` is set as a [Job's trigger](https://trigger.dev/docs/sdk/job) to subscribe to an event a Job from [a sent event](https://trigger.dev/docs/sdk/triggerclient/instancemethods/sendevent)
 * @param options options for the EventTrigger
 */
export function eventTrigger(options) {
    return new EventTrigger({
        name: options.name,
        filter: options.filter,
        source: options.source,
        event: {
            name: options.name,
            title: "Event",
            source: options.source ?? "trigger.dev",
            icon: "custom-event",
            examples: options.examples,
            parsePayload: (rawPayload) => {
                if (options.schema) {
                    const results = options.schema.safeParse(rawPayload);
                    if (!results.success) {
                        throw new ParsedPayloadSchemaError(formatSchemaErrors(results.error.issues));
                    }
                    return results.data;
                }
                return rawPayload;
            },
        },
    });
}
