// @trigger.dev/core/dist/commonjs/v3/limits.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.OFFLOAD_IO_PACKET_LENGTH_LIMIT = exports.OTEL_ATTRIBUTE_PER_EVENT_COUNT_LIMIT = exports.OTEL_ATTRIBUTE_PER_LINK_COUNT_LIMIT = exports.OTEL_LINK_COUNT_LIMIT = exports.OTEL_SPAN_EVENT_COUNT_LIMIT = exports.OTEL_LOG_ATTRIBUTE_VALUE_LENGTH_LIMIT = exports.OTEL_SPAN_ATTRIBUTE_VALUE_LENGTH_LIMIT = exports.OTEL_LOG_ATTRIBUTE_COUNT_LIMIT = exports.OTEL_SPAN_ATTRIBUTE_COUNT_LIMIT = void 0;
exports.imposeAttributeLimits = imposeAttributeLimits;
exports.OTEL_SPAN_ATTRIBUTE_COUNT_LIMIT = 256;
exports.OTEL_LOG_ATTRIBUTE_COUNT_LIMIT = 256;
exports.OTEL_SPAN_ATTRIBUTE_VALUE_LENGTH_LIMIT = 1028;
exports.OTEL_LOG_ATTRIBUTE_VALUE_LENGTH_LIMIT = 1028;
exports.OTEL_SPAN_EVENT_COUNT_LIMIT = 10;
exports.OTEL_LINK_COUNT_LIMIT = 2;
exports.OTEL_ATTRIBUTE_PER_LINK_COUNT_LIMIT = 10;
exports.OTEL_ATTRIBUTE_PER_EVENT_COUNT_LIMIT = 10;
exports.OFFLOAD_IO_PACKET_LENGTH_LIMIT = 128 * 1024;
function imposeAttributeLimits(attributes) {
    const newAttributes = {};
    for (const [key, value] of Object.entries(attributes)) {
        if (calculateAttributeValueLength(value) > exports.OTEL_SPAN_ATTRIBUTE_VALUE_LENGTH_LIMIT) {
            continue;
        }
        if (Object.keys(newAttributes).length >= exports.OTEL_SPAN_ATTRIBUTE_COUNT_LIMIT) {
            break;
        }
        newAttributes[key] = value;
    }
    return newAttributes;
}
function calculateAttributeValueLength(value) {
    if (value === undefined || value === null) {
        return 0;
    }
    if (typeof value === "string") {
        return value.length;
    }
    if (typeof value === "number") {
        return 8;
    }
    if (typeof value === "boolean") {
        return 4;
    }
    if (Array.isArray(value)) {
        return value.reduce((acc, v) => acc + calculateAttributeValueLength(v), 0);
    }
    return 0;
}
